maps-compose/src/main/java/com/google/maps/android/compose/MapViewLifecycleController.kt

```diff
@@ -0,0 +1,136 @@
+        Up, Down
+    }
+
+    private var previousLifecycleState = if (isMapReused)
+        Event.ON_STOP else
+        Event.ON_CREATE
+
+    var lifecycle: Lifecycle? = null
+        set(value) {
+            if (field !== value) {
+                field?.removeObserver(observer)
+                value?.addObserver(observer)
+                field = value
+            }
+        }
+
+    /**
+     * Move to the lifecycle event instead of setting it directly.
+     * */
+    private fun moveToLifecycleEvent(targetLifecycle: Event) {
+        if (targetLifecycle == previousLifecycleState) return
+
+        val lifecycleDirection = if (targetLifecycle in lifecycleUp) Up else Down
+
+        (if (lifecycleDirection == Up) "UP" else "DOWN").let { strDirection ->
+            Log.d(lcTag, "Moving $strDirection from $previousLifecycleState to $targetLifecycle.")
+        }
+
+        do {
+            val nextLifecycleState = nextLifecycleEvent(lifecycleDirection)
+            setLifecycleEvent(nextLifecycleState)
+        } while (nextLifecycleState != targetLifecycle)
+    }
+
+    private fun nextLifecycleEvent(direction: LifecycleDirection) = when (previousLifecycleState) {
+        Event.ON_CREATE -> when (direction) {
+            Up -> Event.ON_START
+            Down -> error("No lifecycle event below ON_CREATE.")
+        }
+
+        Event.ON_START -> when (direction) {
+            Up -> Event.ON_RESUME
+            Down -> error("No lifecycle event below ON_START.")
+        }
+
+        Event.ON_RESUME -> when (direction) {
+            Up -> error("No lifecycle event above ON_RESUME.")
+            Down -> Event.ON_PAUSE
+        }
+
+        Event.ON_PAUSE -> when (direction) {
+            Up -> Event.ON_RESUME
+            Down -> Event.ON_STOP
+        }
+
+        Event.ON_STOP -> when (direction) {
+            Up -> Event.ON_START
+            Down -> Event.ON_DESTROY
+        }
+
+        Event.ON_DESTROY -> error("No lifecycle event above ON_DESTROY")
+        Event.ON_ANY -> error("Unsupported operation")
+    }
+
+
+    private fun setLifecycleEvent(event: Event) {
+        Log.d(lcTag, "Invoking: $event!")
+
+        when (event) {
+            Event.ON_CREATE -> {
+                // Skip calling mapView.onCreate if the lifecycle did not go through onDestroy - in
+                // this case the GoogleMap composable also doesn't leave the composition. So,
+                // recreating the map does not restore state properly which must be avoided.
+                if (previousLifecycleState != Event.ON_STOP) {
+                    mapView.onCreate(Bundle())
+                }
+            }
+
+            Event.ON_START -> mapView.onStart()
+            Event.ON_RESUME -> mapView.onResume()
+            Event.ON_PAUSE -> mapView.onPause()
+            Event.ON_STOP -> mapView.onStop()
+            Event.ON_DESTROY -> mapView.onDestroy()
+
+            else -> throw IllegalStateException()
```

Comment:
Make this `error()` as well for consistency with code above.

Reply:
Sure